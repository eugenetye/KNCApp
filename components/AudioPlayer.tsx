import { View, Text, Pressable } from 'react-native'
import Ionicons from "@expo/vector-icons/Ionicons";
import React from 'react';
import { ProgressBar } from 'react-native-paper';
import { AVPlaybackStatus, Audio } from 'expo-av';

const _onPlaybackUpdate = (status: AVPlaybackStatus, setprog: any) => {
  if (status.isLoaded) {
    const prog = status.durationMillis
      ? status.positionMillis / status.durationMillis
      : 0;

    setprog(prog);
  } else {
    // There was an error...
  }
}

const Controls = ({ playing, onPlay, onPause }: { playing: Boolean, onPlay: () => void, onPause: () => void }) => {
  return (
    <View>
      <View style={{ display: 'flex', flexDirection: 'row' }}>
        {playing ? (
          // Show pause button when playing
          <Pressable onPress={onPause}>
            <Ionicons
              name='pause-circle'
              size={50}
              color={"gray"}
            />
          </Pressable>
        ) : (
          // Show play button when paused
          <Pressable onPress={onPlay}>
            <Ionicons
              name='play-circle'
              size={50}
              color={"#C5DFC5"}
            />
          </Pressable>
        )}
      </View>
    </View>
  );
};

type AudioProps = {
  file: string
};
type AudioState = {
  playbackObj: Audio.Sound | null,
  isPlaying: Boolean,
  hasStarted: Boolean,
  progress: number,
  duration: number,
};


export class AudioPlayer extends React.Component<AudioProps, AudioState> {
  state: AudioState = {
    playbackObj: null,
    isPlaying: false,
    hasStarted: false,
    progress: 0,
    duration: 0,
  };

  async componentDidMount(): Promise<void> {
    const playbackObject = new Audio.Sound();
    playbackObject.setOnPlaybackStatusUpdate(
      (s: AVPlaybackStatus) => {
        _onPlaybackUpdate(
          s,
          (prog: number) => { this.setState({ progress: prog }); },
        );
      }
    );

    try {
      await playbackObject.loadAsync({ uri: this.props.file }, { shouldPlay: false });
    } catch (e) {
      console.log(`LOAD: ${e}`);
      throw e;
      return;
    }
    this.setState({
      playbackObj: playbackObject,
      isPlaying: false,
      hasStarted: true
    });
  }

  async componentDidUpdate(
    prevProps: Readonly<AudioProps>,
    prevState: Readonly<AudioState>,
    snapshot?: any
  ): Promise<void> {
    try {
      if (this.state.hasStarted && this.state.playbackObj) {
        if (prevState.isPlaying && !this.state.isPlaying) {
          await this.state.playbackObj.pauseAsync();
        } else if (!prevState.isPlaying && this.state.isPlaying) {
          await this.state.playbackObj.playAsync();
        }
      }
    } catch (e) {
      console.log(`PLAY/PAUSE: ${e}`);
    }
  }

  async componentWillUnmount(): Promise<void> {
      if (this.state.playbackObj) {
        await this.state.playbackObj.stopAsync();
        await this.state.playbackObj.unloadAsync();
      }
  }

  render(): React.ReactNode {
    return (
      <View style={{
        backgroundColor: "#F8F8F8",
        borderRadius: 8,
        padding: 15,
      }}>
        <Text style={{ padding: 3}}>Listen to the audio file here:</Text>
        <View style={{ height: 60, width: '100%'}}>
          <Controls
            playing={this.state.isPlaying}
            onPlay={() => { this.setState({ isPlaying: true }); }}
            onPause={() => { this.setState({ isPlaying: false }); }}
          />
        </View>
        <View>
          <ProgressBar progress={this.state.progress} />
        </View>
      </View>
    );
  }
}
