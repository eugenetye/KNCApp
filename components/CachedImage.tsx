import { Image } from 'react-native'
import React, { useEffect, useState } from 'react'
import * as FileSystem from 'expo-file-system';
import { FIREBASE_STORAGE } from "../firebaseConfig";
import { ref, getDownloadURL } from "firebase/storage";

export const CachedImage = (props: any) => {
  const { url, style } = props;
  const [cachedUri, setCachedUri] = useState<string | any>(null);

  useEffect(() => {
    Cached();
  }, [])

  const Cached = async () => {
    // Create a file path by concatenating 'FileSystem.cacheDirectory' and 'kncapp/'.
    const path = FileSystem.cacheDirectory + 'kncapp/';

    const category = url.split('/').slice(1, -1).join('/');

    // Ensure the cache directory exists
    await FileSystem.makeDirectoryAsync(path + category + '/', { intermediates: true });

    // Check if the image is already cached
    const cachedImageInfo = await FileSystem.getInfoAsync(`${FileSystem.cacheDirectory}kncapp` + url);

    if (cachedImageInfo.exists) {
      setCachedUri(cachedImageInfo.uri);
    } else {
      const reference = ref(FIREBASE_STORAGE, url);
      try {
        const downloadURL = await getDownloadURL(reference);
        // Download the image and save it to the cache
        const downloadedImage = await FileSystem.downloadAsync(downloadURL, `${FileSystem.cacheDirectory}kncapp` + url);
        setCachedUri(downloadedImage.uri);
      } catch (error) {
        console.error('Error downloading and caching image:', error);
      }
    }
  }

  return (
    <Image source={{ uri: cachedUri }} style={style} />
  )
}


