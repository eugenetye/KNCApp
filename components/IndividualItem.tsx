import { Text, View, Pressable, Image } from "react-native";
import React, { useState, useEffect } from "react";
import { FIREBASE_STORAGE } from "../firebaseConfig";
import { ref, getDownloadURL } from "firebase/storage";
import { useNavigation } from "@react-navigation/native";
import { StackNavigationProp } from '@react-navigation/stack';


export type RootStackParamList = {
  Trail_Info: { param: any } | undefined;
};

export const IndividualItem = ({ item }: any) => {
  const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();

  const imgLink = '/' + item.uid + '.jpeg';

  const [url, setUrl] = useState(imgLink);

  useEffect(() => {
    const func = async () => {
      const reference = ref(FIREBASE_STORAGE, imgLink);
      await getDownloadURL(reference).then((x) => {
        setUrl(x);
      })
    }

    func();
  }, []);

  return (
    <View>
      <Pressable onPress={() => navigation.navigate('Trail_Info', { param: item })}

        style={{
          backgroundColor: "#F8F8F8",
          borderRadius: 8,
          padding: 10,
          flexDirection: "column",
          alignItems: "flex-start",
          justifyContent: "space-between",
          margin: 14,
        }}
      >
        <View style={{ paddingVertical: 12 }}>
          <Image
            style={{ width: 340, height: 150, borderRadius: 8 }}
            source={{ uri: url }}
          />
        </View>

        <View>
          <Text
            style={{
              fontSize: 19,
              fontWeight: "500",
              paddingVertical: 8,

            }}
          >
            {item.name}
          </Text>
        </View>
      </Pressable>
    </View>
  );
};

