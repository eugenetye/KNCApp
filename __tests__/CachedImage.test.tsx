import React from 'react';
import {act, create, ReactTestRendererJSON} from 'react-test-renderer';

import { CachedImage } from '../components/CachedImage';

describe('<CachedImage />', () => {
  // Test Case ID = 5
  it('does not play on invalid song', async () => {
    const comp = create(<CachedImage url='/trails/arboretum_loop.jpeg' />);
    let tree = comp.toTree();
    expect(tree?.props.url).toEqual("/trails/arboretum_loop.jpeg")
  });

  // Test Case ID = 6
  it('renders correctly', async () => {
    const comp = create(<CachedImage url='/trails/arboretum_loop.jpeg' />);
    let tree = comp.toJSON();
    expect(tree).toMatchSnapshot()
  });
});
