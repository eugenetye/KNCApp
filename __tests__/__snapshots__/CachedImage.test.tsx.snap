// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CachedImage /> renders correctly 1`] = `
<Image
  source={
    {
      "uri": null,
    }
  }
/>
`;
